draft-ietf-quic-recovery.md

```diff
@@ -913,32 +913,43 @@ Pseudocode for DetectLostPackets follows:
 DetectLostPackets(largest_acked):
   loss_time = 0
   lost_packets = {}
-  delay_until_lost = infinite
+  loss_delay = 0
+  lost_packet_send_time = now()
+
   if (kUsingTimeLossDetection):
-    delay_until_lost =
-      (1 + time_reordering_fraction) *
-          max(latest_rtt, smoothed_rtt)
+    loss_delay = (1 + time_reordering_fraction) *
+                     max(latest_rtt, smoothed_rtt)
   else if (largest_acked.packet_number == largest_sent_packet):
     // Early retransmit timer.
-    delay_until_lost = 9/8 * max(latest_rtt, smoothed_rtt)
-  foreach (unacked < largest_acked.packet_number):
-    time_since_sent = now() - unacked.time_sent
-    delta = largest_acked.packet_number - unacked.packet_number
-    if (time_since_sent > delay_until_lost ||
-        delta > reordering_threshold):
+    loss_delay = 9/8 * max(latest_rtt, smoothed_rtt)
+  // Packets sent before this time are deemed lost.
+  lost_send_time -= loss_delay
+  lost_pn = largest_acked.packet_number - reordering_threshold
+
+  foreach unacked:
+    if unacked.packet_number > largest_acked.packet_number:
+      continue
+
+    if (unacked.time_sent > lost_send_time ||
+        unacked.packet_number > lost_pn):
       sent_packets.remove(unacked.packet_number)
       if (unacked.retransmittable):
         lost_packets.insert(unacked)
-    else if (loss_time == 0 && delay_until_lost != infinite):
-      loss_time = now() + delay_until_lost - time_since_sent
+    else:
+      loss_time = min(loss_time, unacked.time_sent + loss_delay)
 
   // Inform the congestion controller of lost packets and
   // lets it decide whether to retransmit immediately.
   if (!lost_packets.empty()):
     OnPacketsLost(lost_packets)
 ~~~
 
+This algorithm results in loss_time being set to the earliest time that the
+earliest packet was sent.  As a result loss_time could be in the past.  Timers
```

Comment:
This algorithm results in loss_time being set based on the earliest packet that is still in flight. As a result ...